generate/genfilehcl/genfilehcl.go

```diff
@@ -0,0 +1,287 @@
+type StackHCLs struct {
+	hcls map[string]HCL
+}
+
+// HCL represents generated HCL code from a single block.
+// Is contains parsed and evaluated code on it and information
+// about the origin of the generated code.
+type HCL struct {
+	origin string
+	body   []byte
+}
+
+const (
+	ErrMultiLevelConflict errutil.Error = "conflicting generate_file blocks"
+	ErrParsing            errutil.Error = "parsing generate_file block"
+	ErrEval               errutil.Error = "evaluating generate_file block"
+)
+
+// GeneratedHCLs returns all generated code, mapping the name to its
+// equivalent generated code.
+func (s StackHCLs) GeneratedHCLs() map[string]HCL {
+	cp := map[string]HCL{}
+	for k, v := range s.hcls {
+		cp[k] = v
+	}
+	return cp
+}
+
+// String returns a string representation of the HCL code
+// or an empty string if the config itself is empty.
+func (b HCL) String() string {
+	return string(b.body)
+}
+
+// Origin returns the path, relative to the project root,
+// of the configuration that originated the code.
+func (b HCL) Origin() string {
+	return b.origin
+}
+
+// Load loads from the file system all generate_file for
+// a given stack. It will navigate the file system from the stack dir until
+// it reaches rootdir, loading generate_file and merging them appropriately.
+//
+// More specific definitions (closer or at the stack) have precedence over
+// less specific ones (closer or at the root dir).
+//
+// Metadata and globals for the stack are used on the evaluation of the
+// generate_file blocks.
+//
+// The returned result only contains evaluated values.
+//
+// The rootdir MUST be an absolute path.
+func Load(rootdir string, sm stack.Metadata, globals terramate.Globals) (StackHCLs, error) {
+	stackpath := filepath.Join(rootdir, sm.Path)
+	logger := log.With().
+		Str("action", "genfile.Load()").
+		Str("path", stackpath).
+		Logger()
+
+	logger.Trace().Msg("loading generate_file blocks.")
+
+	loadedFileBlocks, err := loadGenFileBlocks(rootdir, stackpath)
+	if err != nil {
+		return StackHCLs{}, fmt.Errorf("loading generate_file: %w", err)
+	}
+
+	evalctx, err := newEvalCtx(stackpath, sm, globals)
+	if err != nil {
+		return StackHCLs{}, fmt.Errorf("%w: creating eval context: %v", ErrEval, err)
+	}
+
+	logger.Trace().Msg("generating HCL code.")
+
+	res := StackHCLs{
+		hcls: map[string]HCL{},
+	}
+
+	for name, loadedFileBlock := range loadedFileBlocks {
+		logger := logger.With().
+			Str("block", name).
+			Logger()
+
+		logger.Trace().Msg("evaluating block.")
+
+		value, err := evalctx.Eval(loadedFileBlock.content.Expr)
+		if err != nil {
+			evalErr := fmt.Errorf(
+				"%w: stack %q block %q",
+				ErrEval,
+				stackpath,
+				name,
+			)
+			return StackHCLs{}, errutil.Chain(evalErr, err)
+		}
+
+		val := ""
+
+		if !value.IsNull() {
+			val = value.AsString()
+		}
+
+		res.hcls[name] = HCL{
+			origin: loadedFileBlock.origin,
+			body:   hclwrite.Format([]byte(val)),
```

Comment:
Why are we formatting the value as HCL if the idea is to generate arbitrary data ?(YAML/JSON/etc)